In [3]:
import pandas as pd

In [4]:
df1 = pd.read_csv('election_Pak/Kaggle v2 - Election Data.csv')

In [82]:
df2 = pd.read_csv('election_Pak/Kaggle v3 - Election Data.csv')

In [83]:
df3 = pd.read_csv('election_Pak/Kaggle v4 - Election Data.csv')

In [84]:
df1.head()

S.No         Seat Constituency_title                  Candidate_Name  \
0     0   Peshawar I               NA-1         Muhammad Yousaf Khattak   
1     1   Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan   
2     2   Peshawar I               NA-1                       Jan Akbar   
3     3  Peshawar II               NA-2    Arbab Mohammad Jehandir Khan   
4     4  Peshawar II               NA-2            Arbab Saif-ur-Rahman   

    Party  Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0  PML(Q)  44507             79847                 2688       82535   
1     PNA  33964             79847                 2688       82535   
2     IND   1376             79847                 2688       82535   
3     PPP  32247             65825                 2155       67980   
4     PNA  30268             65825                 2155       67980   

  Total_Registered_Voters Turnout  Election  
0                  150063    0.55      1977  
1                  150063    0.55      1977  
2                  150063    0.55      1977  
3                  129733   0.524      1977  
4                  129733   0.524      1977

In [85]:
df2.head()

Unnamed: 0         Seat Constituency_title                  Candidate_Name  \
0           0   Peshawar I               NA-1         Muhammad Yousaf Khattak   
1           1   Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan   
2           2   Peshawar I               NA-1                       Jan Akbar   
3           3  Peshawar II               NA-2    Arbab Mohammad Jehandir Khan   
4           4  Peshawar II               NA-2            Arbab Saif-ur-Rahman   

    Party  Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0  PML(Q)  44507             79847                 2688       82535   
1     PNA  33964             79847                 2688       82535   
2     IND   1376             79847                 2688       82535   
3     PPP  32247             65825                 2155       67980   
4     PNA  30268             65825                 2155       67980   

  Total_Registered_Voters Turnout  Election  
0                  150063    0.55      1977  
1                  150063    0.55      1977  
2                  150063    0.55      1977  
3                  129733   0.524      1977  
4                  129733   0.524      1977

In [86]:
df3.head()

Unnamed: 0         Seat Constituency_title                  Candidate_Name  \
0           0   Peshawar-I               NA-1         Muhammad Yousaf Khattak   
1           1   Peshawar-I               NA-1  Air Marshal (Retd) Asghar Khan   
2           2   Peshawar-I               NA-1                       Jan Akbar   
3           3  Peshawar-II               NA-2    Arbab Mohammad Jehandir Khan   
4           4  Peshawar-II               NA-2            Arbab Saif-ur-Rahman   

    Party  Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0  PML(Q)  44507             79847                 2688       82535   
1     PNA  33964             79847                 2688       82535   
2     IND   1376             79847                 2688       82535   
3     PPP  32247             65825                 2155       67980   
4     PNA  30268             65825                 2155       67980   

  Total_Registered_Voters Turnout  Election  
0                  150063    0.55      1977  
1                  150063    0.55      1977  
2                  150063    0.55      1977  
3                  129733   0.524      1977  
4                  129733   0.524      1977

In [87]:
df3.tail()

Unnamed: 0                Seat Constituency_title  \
19712       15212  Lasbela-cum-Gwadar             NA-272   
19713       15213  Lasbela-cum-Gwadar             NA-272   
19714       15214  Lasbela-cum-Gwadar             NA-272   
19715       15215  Lasbela-cum-Gwadar             NA-272   
19716       15216  Lasbela-cum-Gwadar             NA-272   

               Candidate_Name                       Party  Votes  \
19712          Mohammad Akram         Pak Sarzameen Party    882   
19713  Muhammad Akhtar mengal  Balochistan National Party  41866   
19714      Nawab Khan Bizenjo          Amun Taraqqi Party    162   
19715           Shakeel Ahmed                 Independent    338   
19716     Wazeer Ahmed Norani  Tehreek-e-Labbaik Pakistan   4604   

      Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
19712                                                      
19713                                                      
19714                                                      
19715                                                      
19716                                                      

      Total_Registered_Voters Turnout  Election  
19712                          56.02%      2018  
19713                          56.02%      2018  
19714                          56.02%      2018  
19715                          56.02%      2018  
19716                          56.02%      2018

In [88]:
df1.drop('S.No', axis=1, inplace=True)
df2.drop('Unnamed: 0', axis=1, inplace=True)
df3.drop('Unnamed: 0', axis=1, inplace=True)

In [89]:
df1.head(3)

Seat Constituency_title                  Candidate_Name   Party  \
0  Peshawar I               NA-1         Muhammad Yousaf Khattak  PML(Q)   
1  Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan     PNA   
2  Peshawar I               NA-1                       Jan Akbar     IND   

   Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0  44507             79847                 2688       82535   
1  33964             79847                 2688       82535   
2   1376             79847                 2688       82535   

  Total_Registered_Voters Turnout  Election  
0                  150063    0.55      1977  
1                  150063    0.55      1977  
2                  150063    0.55      1977

In [90]:
df2.head(3)

Seat Constituency_title                  Candidate_Name   Party  \
0  Peshawar I               NA-1         Muhammad Yousaf Khattak  PML(Q)   
1  Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan     PNA   
2  Peshawar I               NA-1                       Jan Akbar     IND   

   Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0  44507             79847                 2688       82535   
1  33964             79847                 2688       82535   
2   1376             79847                 2688       82535   

  Total_Registered_Voters Turnout  Election  
0                  150063    0.55      1977  
1                  150063    0.55      1977  
2                  150063    0.55      1977

In [91]:
df = pd.concat([df1,df2])

In [92]:
df

Seat Constituency_title                  Candidate_Name  \
0               Peshawar I               NA-1         Muhammad Yousaf Khattak   
1               Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan   
2               Peshawar I               NA-1                       Jan Akbar   
3              Peshawar II               NA-2    Arbab Mohammad Jehandir Khan   
4              Peshawar II               NA-2            Arbab Saif-ur-Rahman   
...                    ...                ...                             ...   
19722  Lasbela-cum-Gawadar             NA 206         Molvi M. Yaqoob Sasoli    
19723  Lasbela-cum-Gawadar             NA 206                   Abdul Hameed    
19724   Turbat-cum-Panjgur             NA 207     Shakeel Ahmad Baloch, Adv.    
19725   Turbat-cum-Panjgur             NA 207            Dr. M. Yasin Baloch    
19726   Turbat-cum-Panjgur             NA 207              Rehmatullah Molvi    

          Party  Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0        PML(Q)  44507             79847                 2688       82535   
1           PNA  33964             79847                 2688       82535   
2           IND   1376             79847                 2688       82535   
3           PPP  32247             65825                 2155       67980   
4           PNA  30268             65825                 2155       67980   
...         ...    ...               ...                  ...         ...   
19722      IND    1061             58265                 2918       61183   
19723      IND     926             58265                 2918       61183   
19724      BNP   39838             73000                  245       73245   
19725      BNM   27863             73000                  245       73245   
19726  JUI (F)    5299             73000                  245       73245   

      Total_Registered_Voters Turnout  Election  
0                      150063    0.55      1977  
1                      150063    0.55      1977  
2                      150063    0.55      1977  
3                      129733   0.524      1977  
4                      129733   0.524      1977  
...                       ...     ...       ...  
19722                  199228   0.307      1997  
19723                  199228   0.307      1997  
19724                  243554   0.301      1997  
19725                  243554   0.301      1997  
19726                  243554   0.301      1997  

[39454 rows x 11 columns]

In [93]:
df = pd.concat([df,df3])

In [94]:
df.head(10)

Seat Constituency_title                   Candidate_Name    Party  \
0    Peshawar I               NA-1          Muhammad Yousaf Khattak   PML(Q)   
1    Peshawar I               NA-1   Air Marshal (Retd) Asghar Khan      PNA   
2    Peshawar I               NA-1                        Jan Akbar      IND   
3   Peshawar II               NA-2     Arbab Mohammad Jehandir Khan      PPP   
4   Peshawar II               NA-2             Arbab Saif-ur-Rahman      PNA   
5   Peshawar II               NA-2                    M. Nazirullah  PML (Q)   
6   Peshawar II               NA-2  Sqr. Leader (R) Qazi Shah Jehan      IND   
7   Peshawar II               NA-2                    M. Aslam Khan      IND   
8  Peshawar III               NA-3                 Aftab Ahmed Khan      PPP   
9  Peshawar III               NA-3       Sardar Sherbaz Khan Mazari      PNA   

   Votes Total_Valid_Votes Total_Rejected_Votes Total_Votes  \
0  44507             79847                 2688       82535   
1  33964             79847                 2688       82535   
2   1376             79847                 2688       82535   
3  32247             65825                 2155       67980   
4  30268             65825                 2155       67980   
5   2034             65825                 2155       67980   
6    674             65825                 2155       67980   
7    602             65825                 2155       67980   
8  40917             75728                 2546       78274   
9  32216             75728                 2546       78274   

  Total_Registered_Voters Turnout  Election  
0                  150063    0.55      1977  
1                  150063    0.55      1977  
2                  150063    0.55      1977  
3                  129733   0.524      1977  
4                  129733   0.524      1977  
5                  129733   0.524      1977  
6                  129733   0.524      1977  
7                  129733   0.524      1977  
8                  135188   0.579      1977  
9                  135188   0.579      1977

In [95]:
print("length of df1 = ", len(df1))
print("length of df2 = ", len(df2))
print("length of df3 = ", len(df3))

if (len(df1)+len(df2)+len(df3)) == len(df):
    print("All records concatenated")
else:
    print("Not All records concatenated")

length of df1 =  19727
length of df2 =  19727
length of df3 =  19717
All records concatenated


In [96]:
df.isnull().sum()

Seat                       6876
Constituency_title            0
Candidate_Name               12
Party                      3318
Votes                      7609
Total_Valid_Votes          7791
Total_Rejected_Votes       7791
Total_Votes                8800
Total_Registered_Voters    7117
Turnout                    7806
Election                      0
dtype: int64

In [97]:
len(df)

59171

In [98]:
(59171 * 20)/100 

11834.2

So if we have more than 11800 NaN values we will be removing those rows because it will skew over data, but according to the nature of this problem, we will be asking business domain experts, how to treat with this data!

BUT WAIT ! ! ! 

maybe it is not counting the empty cells, but only NaNs

In [99]:
import numpy as np

In [100]:
df.replace(' ', np.nan, inplace=True)

In [101]:
df

Seat Constituency_title                  Candidate_Name  \
0              Peshawar I               NA-1         Muhammad Yousaf Khattak   
1              Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan   
2              Peshawar I               NA-1                       Jan Akbar   
3             Peshawar II               NA-2    Arbab Mohammad Jehandir Khan   
4             Peshawar II               NA-2            Arbab Saif-ur-Rahman   
...                   ...                ...                             ...   
19712  Lasbela-cum-Gwadar             NA-272                  Mohammad Akram   
19713  Lasbela-cum-Gwadar             NA-272          Muhammad Akhtar mengal   
19714  Lasbela-cum-Gwadar             NA-272              Nawab Khan Bizenjo   
19715  Lasbela-cum-Gwadar             NA-272                   Shakeel Ahmed   
19716  Lasbela-cum-Gwadar             NA-272             Wazeer Ahmed Norani   

                            Party  Votes Total_Valid_Votes  \
0                          PML(Q)  44507             79847   
1                             PNA  33964             79847   
2                             IND   1376             79847   
3                             PPP  32247             65825   
4                             PNA  30268             65825   
...                           ...    ...               ...   
19712         Pak Sarzameen Party    882               NaN   
19713  Balochistan National Party  41866               NaN   
19714          Amun Taraqqi Party    162               NaN   
19715                 Independent    338               NaN   
19716  Tehreek-e-Labbaik Pakistan   4604               NaN   

      Total_Rejected_Votes Total_Votes Total_Registered_Voters Turnout  \
0                     2688       82535                  150063    0.55   
1                     2688       82535                  150063    0.55   
2                     2688       82535                  150063    0.55   
3                     2155       67980                  129733   0.524   
4                     2155       67980                  129733   0.524   
...                    ...         ...                     ...     ...   
19712                  NaN         NaN                     NaN  56.02%   
19713                  NaN         NaN                     NaN  56.02%   
19714                  NaN         NaN                     NaN  56.02%   
19715                  NaN         NaN                     NaN  56.02%   
19716                  NaN         NaN                     NaN  56.02%   

       Election  
0          1977  
1          1977  
2          1977  
3          1977  
4          1977  
...         ...  
19712      2018  
19713      2018  
19714      2018  
19715      2018  
19716      2018  

[59171 rows x 11 columns]

In [102]:
df.isnull().sum()

Seat                        6876
Constituency_title             0
Candidate_Name                12
Party                       3318
Votes                       7609
Total_Valid_Votes          10940
Total_Rejected_Votes       10941
Total_Votes                11948
Total_Registered_Voters    10265
Turnout                     7806
Election                       0
dtype: int64

Now, re-calculate the equation

In [103]:
len(df)

59171

In [104]:
(59171 * 20)/100

11834.2

Now, according to this we'll need to delete Total_Votes columns OR we'll be increasing the our threshold amount, like above than 20% or 25%. Let's stick to our policy! 20%

In [105]:
df.drop('Total_Votes', axis=1, inplace=True)

In [106]:
df.replace(' NA', np.nan, inplace=True)

In [107]:
df.isnull().sum()

Seat                        6876
Constituency_title             0
Candidate_Name                12
Party                       3318
Votes                       7631
Total_Valid_Votes          10940
Total_Rejected_Votes       10941
Total_Registered_Voters    10265
Turnout                     7806
Election                       0
dtype: int64

Just notice Votes in this output & in previous output.

In [108]:
df.replace(' NA ', np.nan, inplace=True)

In [109]:
df.isnull().sum()

Seat                        6876
Constituency_title             0
Candidate_Name                12
Party                       3318
Votes                       7635
Total_Valid_Votes          10940
Total_Rejected_Votes       10941
Total_Registered_Voters    10265
Turnout                     7806
Election                       0
dtype: int64

Just notice Votes in this output & in previous output. 
7631 earlier, now 7635

In [111]:
df.where(df == 'NaN')

/home/umair/anaconda3/lib/python3.7/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


Seat Constituency_title Candidate_Name Party Votes Total_Valid_Votes  \
0      NaN                NaN            NaN   NaN   NaN               NaN   
1      NaN                NaN            NaN   NaN   NaN               NaN   
2      NaN                NaN            NaN   NaN   NaN               NaN   
3      NaN                NaN            NaN   NaN   NaN               NaN   
4      NaN                NaN            NaN   NaN   NaN               NaN   
...    ...                ...            ...   ...   ...               ...   
19712  NaN                NaN            NaN   NaN   NaN               NaN   
19713  NaN                NaN            NaN   NaN   NaN               NaN   
19714  NaN                NaN            NaN   NaN   NaN               NaN   
19715  NaN                NaN            NaN   NaN   NaN               NaN   
19716  NaN                NaN            NaN   NaN   NaN               NaN   

      Total_Rejected_Votes Total_Registered_Voters Turnout  Election  
0                      NaN                     NaN     NaN       NaN  
1                      NaN                     NaN     NaN       NaN  
2                      NaN                     NaN     NaN       NaN  
3                      NaN                     NaN     NaN       NaN  
4                      NaN                     NaN     NaN       NaN  
...                    ...                     ...     ...       ...  
19712                  NaN                     NaN     NaN       NaN  
19713                  NaN                     NaN     NaN       NaN  
19714                  NaN                     NaN     NaN       NaN  
19715                  NaN                     NaN     NaN       NaN  
19716                  NaN                     NaN     NaN       NaN  

[59171 rows x 10 columns]

In [112]:
cols = list(df.columns)

In [113]:
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values = np.nan, strategy='most_frequent')

a = imp.fit_transform(df)

In [114]:
df = pd.DataFrame(a)

In [115]:
df.columns = cols

In [116]:
df.head()

Seat Constituency_title                  Candidate_Name   Party  \
0   Peshawar I               NA-1         Muhammad Yousaf Khattak  PML(Q)   
1   Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan     PNA   
2   Peshawar I               NA-1                       Jan Akbar     IND   
3  Peshawar II               NA-2    Arbab Mohammad Jehandir Khan     PPP   
4  Peshawar II               NA-2            Arbab Saif-ur-Rahman     PNA   

   Votes Total_Valid_Votes Total_Rejected_Votes Total_Registered_Voters  \
0  44507             79847                 2688                  150063   
1  33964             79847                 2688                  150063   
2   1376             79847                 2688                  150063   
3  32247             65825                 2155                  129733   
4  30268             65825                 2155                  129733   

  Turnout Election  
0    0.55     1977  
1    0.55     1977  
2    0.55     1977  
3   0.524     1977  
4   0.524     1977

Now, this below features are important when we have to delete all columns which contain every value as NaN or some
values as NaN, so that we've the names of columns where we can perform dropping.

In [117]:
colsWithAllNullValues = df.columns[df.isnull().all()].tolist()

In [118]:
colsWithAllNullValues

[]

In [119]:
colsWithAllNullValues = df.columns[df.isnull().any()].tolist()

In [120]:
colsWithAllNullValues

[]

In [121]:
len(df)

59171

In [122]:
df.drop_duplicates(subset = cols, keep = 'first', inplace = True)

In [123]:
len(df)

36035

In [124]:
df.columns

Index(['Seat', 'Constituency_title', 'Candidate_Name', 'Party', 'Votes',
       'Total_Valid_Votes', 'Total_Rejected_Votes', 'Total_Registered_Voters',
       'Turnout', 'Election'],
      dtype='object')

In [ ]:
from sklearn.ensemble.RandomForestRegressor

rf = RandomForestRegressor()
rf.fit(X, y)
print (sorted(zip(map(lambda x: round(x, 4), rf.feature_importances_), names), 
             reverse=True))

In [125]:
df

Seat Constituency_title                  Candidate_Name  \
0              Peshawar I               NA-1         Muhammad Yousaf Khattak   
1              Peshawar I               NA-1  Air Marshal (Retd) Asghar Khan   
2              Peshawar I               NA-1                       Jan Akbar   
3             Peshawar II               NA-2    Arbab Mohammad Jehandir Khan   
4             Peshawar II               NA-2            Arbab Saif-ur-Rahman   
...                   ...                ...                             ...   
59166  Lasbela-cum-Gwadar             NA-272                  Mohammad Akram   
59167  Lasbela-cum-Gwadar             NA-272          Muhammad Akhtar mengal   
59168  Lasbela-cum-Gwadar             NA-272              Nawab Khan Bizenjo   
59169  Lasbela-cum-Gwadar             NA-272                   Shakeel Ahmed   
59170  Lasbela-cum-Gwadar             NA-272             Wazeer Ahmed Norani   

                            Party  Votes Total_Valid_Votes  \
0                          PML(Q)  44507             79847   
1                             PNA  33964             79847   
2                             IND   1376             79847   
3                             PPP  32247             65825   
4                             PNA  30268             65825   
...                           ...    ...               ...   
59166         Pak Sarzameen Party    882            126098   
59167  Balochistan National Party  41866            126098   
59168          Amun Taraqqi Party    162            126098   
59169                 Independent    338            126098   
59170  Tehreek-e-Labbaik Pakistan   4604            126098   

      Total_Rejected_Votes Total_Registered_Voters Turnout Election  
0                     2688                  150063    0.55     1977  
1                     2688                  150063    0.55     1977  
2                     2688                  150063    0.55     1977  
3                     2155                  129733   0.524     1977  
4                     2155                  129733   0.524     1977  
...                    ...                     ...     ...      ...  
59166                    0                  312633  56.02%     2018  
59167                    0                  312633  56.02%     2018  
59168                    0                  312633  56.02%     2018  
59169                    0                  312633  56.02%     2018  
59170                    0                  312633  56.02%     2018  

[36035 rows x 10 columns]

In [127]:
df.query('Seat == "Peshawar I" and Constituency_title == "NA-1"', inplace = True)

In [128]:
df

Seat Constituency_title                     Candidate_Name  \
0     Peshawar I               NA-1            Muhammad Yousaf Khattak   
1     Peshawar I               NA-1     Air Marshal (Retd) Asghar Khan   
2     Peshawar I               NA-1                          Jan Akbar   
741   Peshawar I               NA-1         Haji Mohammad Younis Elahi   
742   Peshawar I               NA-1                Zafar Ali Shah Syed   
743   Peshawar I               NA-1                        M. Khurshid   
744   Peshawar I               NA-1  Maj. (R) Sultan Hussain Shah Syed   
1839  Peshawar I               NA-1           Aftab Ahmed Khan Sherpao   
1840  Peshawar I               NA-1                Ghulam Ahmed Bilour   
1841  Peshawar I               NA-1                           M. Rafiq   
1842  Peshawar I               NA-1                         Said Anwar   

            Party  Votes Total_Valid_Votes Total_Rejected_Votes  \
0          PML(Q)  44507             79847                 2688   
1             PNA  33964             79847                 2688   
2             IND   1376             79847                 2688   
741   Independent  26815             55597                 1768   
742   Independent  22524             55597                 1768   
743   Independent   5779             55597                 1768   
744   Independent    479             55597                 1768   
1839          PPP  44658             84766                 1470   
1840          ANP  35947             84766                 1470   
1841          IND   2335             84766                 1470   
1842          IND   1826             84766                 1470   

     Total_Registered_Voters Turnout Election  
0                     150063    0.55     1977  
1                     150063    0.55     1977  
2                     150063    0.55     1977  
741                   168846    0.34     1985  
742                   168846    0.34     1985  
743                   168846    0.34     1985  
744                   168846    0.34     1985  
1839                  226170   0.381     1988  
1840                  226170   0.381     1988  
1841                  226170   0.381     1988  
1842                  226170   0.381     1988

In [7]:
df1.columns[-1]

'Election'

In [8]:
df1[df1.columns[-1]]


0        1977
1        1977
2        1977
3        1977
4        1977
         ... 
19722    1997
19723    1997
19724    1997
19725    1997
19726    1997
Name: Election, Length: 19727, dtype: int64